### PR TITLE
Improve control of outgoing connection lifecycles

### LINKED ISSUE #67873: `node-left ... reason: disconnected` triggered by earlier `node-left` event rather than network issue
If a node leaves the cluster with `reason: disconnected` then this means something closed one of the TCP channels initiated by the master targeting the departing node. Nodes never deliberately close any incoming connections, and the master doesn't spontaneously close any outgoing connections, so we usually consider `reason: disconnected` to be evidence of a network issue.

However, I recently observed a `node-left` event with `reason: disconnected` on a cluster running entirely on `localhost`, which therefore rules out a network issue. This was the second of two `node-left` events for the same node within a few seconds; the first had reason `followers check retry count exceeded`. I have so far failed to reproduce this situation.

My best guess regarding the cause is that we hadn't closed everything by the time the node rejoins, so the `node-join` task executes and only then does the channel get closed and the master notified that this node disconnected.  (`io.netty.channel.AbstractChannelHandlerContext#close()` is fire-and-forget, it just adds a task to be processed by the event loop, but that particular event loop was blocked so the closing was delayed).

I don't think the behaviour is particularly bad, the node already left the cluster, but the bug is in how we report this as a network event when really the network is fine.

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we open connections to other nodes in various places and largely
assume that they remain open as needed, only closing them when applying
a cluster state that removes the remote node from the cluster. This
isn't ideal: we might preserve unnecessary connections to remote nodes
that aren't in the cluster if they never manage to join the cluster, and
we might also disconnect from a node that left the cluster while it's in
the process of re-joining too (see #67873).

With this commit we move to a model in which each user of a connection
to a remote node acquires a reference to the connection that must be
released once it's no longer needed. Connections remain open while there
are any live references, but are now actively closed when all references
are released.

Fixes #67873

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java [hunk 1]
```diff
@@ -243,239 +231,82 @@ public void reconnectToNodes(DiscoveryNodes discoveryNodes, Runnable onCompletio
...
-     * Additionally if the last step of the process was a disconnection then this target is removed from the current set of targets. Thus
-     * if this {@link ConnectionTarget} is idle and in the current set of targets then it should be connected.
-     * <p>
-     * All of the {@code listeners} are awaiting the completion of the same activity, which is either a connection or a disconnection.  If
-     * we are currently connecting and then {@link ConnectionTarget#disconnect()} is called then all connection listeners are
-     * removed from the list so they can be notified of failure; once the connecting process has finished a disconnection will be started.
-     * Similarly if we are currently disconnecting and then {@link ConnectionTarget#connect(ActionListener)} is called then all
-     * disconnection listeners are immediately removed for failure notification and a connection is started once the disconnection is
-     * complete.
-     */
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:280]: Treat the changes to `NodeConnectionsService` as a complete rewrite (with slightly different semantics, see test changes)

**Review by DaveCTurner — COMMENTED** [review]
Note to reviewers: it will be easier to read this as a sequence of commits rather than looking just at the finished result.

File: server/src/main/java/org/elasticsearch/discovery/PeerFinder.java [hunk 2]
```diff
@@ -111,16 +113,36 @@ public void activate(final DiscoveryNodes lastAcceptedNodes) {
...
         if (peersRemoved) {
             onFoundPeersUpdated();
         }
+
+        // Discovery is over, we're joining a cluster, so we can release all the connections that were being used for discovery. We haven't
+        // finished joining/forming the cluster yet, but if we're joining an existing master then the join will hold open the connection
+        // it's using and if we're becoming the master then join validation will hold open the connections to the joining peers; this set of
+        // peers is a quorum so that's good enough.
+        //
+        // Note however that this might still close connections to other master-eligible nodes that we discovered but which aren't currently
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/discovery/PeerFinder.java:134]: This is a change in behaviour vs today. As per the comment I think it's ok, but we can discuss alternatives too.

File: server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java [hunk 3]
```diff
@@ -90,75 +118,116 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
...
+                                    connectedNodes.remove(node, conn);
                                     connectionListener.onNodeDisconnected(node, conn);
+                                    conn.onRemoved();
+                                }));
+
+                                conn.addCloseListener(ActionListener.wrap(() -> {
+                                    if (connectingRefCounter.hasReferences() == false) {
+                                        logger.trace("connection manager shut down, closing transport connection to [{}]", node);
+                                    } else if (conn.hasReferences()) {
+                                        logger.info("transport connection to [{}] closed by remote", node.descriptionWithoutAttributes());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java:205]: This also emits logs for connections to nodes in remote clusters. I think that's useful to see, we investigate cases that turn out to be flaky cross-cluster connections at a nonzero rate.

**Review by henningandersen — COMMENTED** [review]
I have been through most of this, but not all the tests yet. This looks great and I like how it simplifies `NodeConnectionsService`, seems more natural now.

File: server/src/main/java/org/elasticsearch/discovery/PeerFinder.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/discovery/PeerFinder.java:134]: We discussed in another channel and decided that this won't happen: we're only closing connections that we initiated, the changes here won't have any effect on incoming connections.

File: libs/core/src/main/java/org/elasticsearch/core/AbstractRefCounted.java [hunk 4]
```diff
@@ -58,6 +58,11 @@ public final boolean decRef() {
         return false;
     }
 
+    @Override
+    public boolean hasReferences() {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply libs/core/src/main/java/org/elasticsearch/core/AbstractRefCounted.java]: Yes, see 4811a35

File: server/src/test/java/org/elasticsearch/transport/ClusterConnectionManagerTests.java [hunk 5]
```diff
@@ -220,6 +314,67 @@ public void testConcurrentConnects() throws Exception {
...
+                }
+            }
+        };
+
+        for (int i = 0; i < threadCount; i++) {
+            threadPool.generic().execute(action);
+        }
+
+        assertTrue(countDownLatch.await(10, TimeUnit.SECONDS));
+        connectionManager.close();
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/transport/ClusterConnectionManagerTests.java:376]: Yes, see 4811a35

File: server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java [hunk 6]
```diff
@@ -139,22 +136,13 @@ public void disconnectFromNodesExcept(DiscoveryNodes discoveryNodes) {
...
 
             for (final DiscoveryNode discoveryNode : nodesToDisconnect) {
-                runnables.add(targetsByNode.get(discoveryNode).disconnect());
+                runnables.add(targetsByNode.remove(discoveryNode)::disconnect);
             }
         }
         runnables.forEach(Runnable::run);
     }
 
     void ensureConnections(Runnable onCompletion) {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java]: Aha yes you're absolutely right, thanks. I merged them together in c8a6902.

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/internalClusterTest/java/org/elasticsearch/discovery/ClusterDisruptionIT.java [hunk 7]
```diff
@@ -494,4 +502,60 @@ public void testRestartNodeWhileIndexing() throws Exception {
         }
     }
 
+    public void testRejoinWhileBeingRemoved() {
+        final String masterNode = internalCluster().startMasterOnlyNode();
+        final String dataNode = internalCluster().startDataOnlyNode(Settings.builder()
+            .put(DISCOVERY_FIND_PEERS_INTERVAL_SETTING.getKey(), "100ms")
+            .put(LEADER_CHECK_INTERVAL_SETTING.getKey(), "100ms")
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/discovery/ClusterDisruptionIT.java:514]: Ah that's a good point, thanks. Done in 03df026.

**Pull request merged**